### PR TITLE
Enhance LSP

### PR DESCRIPTION
**Pull request by imkiva**
- Highlight and Goto Def. using core terms to support serialized dependencies
- Hover mouse on a term to show its type
- Find references

### CODE REVIEW

File: lsp/src/main/java/org/aya/lsp/server/AyaServer.java [hunk 1]
```diff
@@ -54,6 +54,8 @@
       workOps.setChangeNotifications(true);
       workCap.setWorkspaceFolders(workOps);
       cap.setWorkspace(workCap);
+      cap.setHoverProvider(true);
+      cap.setSignatureHelpProvider(new SignatureHelpOptions(Collections.singletonList(" ")));
```
Comment (ice1000) [inline lsp/src/main/java/org/aya/lsp/server/AyaServer.java:58]: What's this list?

File: lsp/src/main/java/org/aya/lsp/server/AyaService.java [hunk 2]
```diff
@@ -272,6 +273,27 @@ private DiagnosticSeverity severityOf(@NotNull Problem problem) {
     });
   }
 
+  @Override public CompletableFuture<Hover> hover(HoverParams params) {
+    return CompletableFuture.supplyAsync(() -> {
+      var loadedFile = find(params.getTextDocument().getUri());
+      if (loadedFile == null) return null;
+      var doc = ComputeSignature.invoke(loadedFile, params.getPosition(), true);
+      if (doc.isEmpty()) return null;
+      return new Hover(new MarkupContent(MarkupKind.PLAINTEXT, Doc.stickySep(doc).commonRender()));
```
Comment (ice1000) [inline lsp/src/main/java/org/aya/lsp/server/AyaService.java:282]: Remember that `commonRender` sets the width limit to 80. The unlimited version is `debugRender`

File: lsp/src/main/java/org/aya/lsp/utils/Resolver.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  /** resolve a symbol by its qualified name in the whole library */
+  static @NotNull Option<@NotNull Def> resolveDef(
+    @NotNull LibraryOwner owner,
+    @NotNull ImmutableSeq<String> module,
+    @NotNull String name
+  ) {
+    var mod = findModule(owner, module);
+    return mod.mapNotNull(m -> m.tycked().value)
+      .map(defs -> defs.flatMap(Resolver::withSubLevel))
+      .mapNotNull(defs -> defs.find(def -> def.ref().name().equals(name)).getOrNull());
```
Comment (ice1000) [inline lsp/src/main/java/org/aya/lsp/utils/Resolver.java:36]:
```suggestion
      .flatMap(defs -> defs.find(def -> def.ref().name().equals(name)));
```

**Review by ice1000 — COMMENTED** [review]
image

File: lsp/src/main/java/org/aya/lsp/server/AyaServer.java (hunk 1 above)
Comment (imkiva) [reply lsp/src/main/java/org/aya/lsp/server/AyaServer.java:58]: Tell the frontend that we support hover and signature help (like Ctrl+P)

File: lsp/src/main/java/org/aya/lsp/server/AyaService.java (hunk 2 above)
Comment (imkiva) [reply lsp/src/main/java/org/aya/lsp/server/AyaService.java:282]: I forgot that 🤣

File: lsp/src/main/java/org/aya/lsp/utils/Resolver.java (hunk 3 above)
Comment (imkiva) [reply lsp/src/main/java/org/aya/lsp/utils/Resolver.java:36]: 那你能救救我吗

Comment (ice1000) [reply lsp/src/main/java/org/aya/lsp/utils/Resolver.java:36]: 一点点微小的改动，不值一提，你提交了吧（

**Review by ice1000 — APPROVED** [review]
Merge?

**Comment by imkiva** [comment]
> Merge?

Let me try signature help one more time 😭

**Comment by imkiva** [comment]
bors d=ice1000

**Comment by imkiva** [comment]
bors merge

**Pull request merged**